orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -869,89 +869,257 @@ common::Status HandleSharedInitializer(Graph& graph,
+//   1. We collect all tensors that need be sent from stage to stage.
+//   2. We collect all tensors that need to be forwarded, i.e., tensors that
+// need to be sent by a stage even though it wasn't produced in that stage.
+//   3. We create replicas of all the tensors that will be sent.
+//   4. Finally, for each pair of stages (origin, destination), we perform the
+// following steps:
+//     a) Create all the meta tensors required (e.g., signals, and source and
+// destination ranks).
+//     b) Add as input of the Send operator, all the tensors collected in step 1,
+// and as output of the respective Receive operator.
+//     c) If origin, needs to send any received tensor (a forwarded tensor),
+// mark the replica of that tensor as input and output of the same Send and
+// Receive operators, respectively.
+//     d) Update the consumers of the received tensors to read from the replicas
+// of such tensors.
+//     e) Finally, create the Send and Receive nodes.
+common::Status SplitGraphWithOperatorToStageMap(Graph& graph,
+                                                std::map<Node*, int>& op_to_stage,
+                                                int num_stages,
+                                                std::vector<std::pair<int, int>>& messages,
+                                                std::vector<Node*>& send_nodes,
+                                                std::vector<Node*>& receive_nodes,
+                                                std::vector<int>& stage_to_rank) {
+  
+  // forward_messages stores all the tensors that will be sent by any stage.
+  // For example, forward_messages[s] is a set of all the tensors sent by stage
+  // s while executing the forward computation.
+  std::vector<std::set<const NodeArg*>> forward_messages(num_stages);
+  // TODO(jufranc): once we start using this function on the training graph,
+  // we need to keep backward_messages[s] too.
+
+  // Tensors that need to be sent from one device to the other. For example,
+  // forwarded_tensors[i] = {t, {stage of producer - s0, stage of the last consumer - s1}}
+  // means that a tensor t, is produced in stage s0 and consumed for the last
+  // time in stage s1, where s1 > s0 + 1. This tensor will be copied from stage
+  // to stage until s1.
+  // TODO: Should we consider weights here too?
+  std::vector<std::pair<const NodeArg*, std::pair<int, int>>> forwarded_tensors;
+
+  // All the tensors that are produced and consumed in the graph.
+  auto initial_node_args = GetAllNodeArgs(graph);
+
+  // We create all the tensor replicas in advance using this fuction.
+  // A tensor produced in stage r and consumed in stage r', such that r' > r,
+  // will have a replica in all stages r'', such that r'' > r and r'' < r'.
+  // tensor_replicas[t][r] contains a pointer to the the replica of t in stage r,
+  // it if exists, or to itself if r is the stage of the producer of r.
```

Comment:
>it if [](start = 5, length = 5)

it if -> if it. 

Question: if tensor_replicas[t][r] is null, it means the tensor t doesn't exist.

Reply:
yes. For example, if t is produced in stage 2 and consumed in stage 3, in a pipeline of 4 stages, then tensor_replicas[t][0] and tensor_replicas[t][1] are null.